File: gcs/ui/camera_view.py
"""Live webcam view — a second full-screen-capable view alongside the map.

Uses Qt Multimedia (``QCamera`` → ``QMediaCaptureSession`` → ``QVideoWidget``)
so it works on Windows / Linux with whatever cameras the OS exposes. The view is
self-contained: a header strip picks the device and starts/stops the feed, and a
placeholder is shown until a frame is flowing. In picture-in-picture mode the
header is hidden so only the live image (or placeholder) fills the small tile.

It is deliberately passive about layout — the MainWindow decides whether it is
the primary view or the corner PIP and calls :meth:`set_pip` accordingly.
"""
from __future__ import annotations

from PySide6.QtCore import Qt, Signal
from PySide6.QtWidgets import (QComboBox, QHBoxLayout, QLabel, QPushButton,
                               QStackedLayout, QVBoxLayout, QWidget)

try:
    from PySide6.QtMultimedia import (QCamera, QMediaCaptureSession,
                                      QMediaDevices)
    from PySide6.QtMultimediaWidgets import QVideoWidget
    _MULTIMEDIA_OK = True
except Exception:  # pragma: no cover - depends on the PySide6 build
    _MULTIMEDIA_OK = False


class CameraView(QWidget):
    """A webcam panel that can act as the main view or shrink to a PIP tile."""

    running_changed = Signal(bool)

    def __init__(self, parent=None):
        super().__init__(parent)
        self._running = False
        self._camera = None
        self._media_devices = None
        self._session = None
        self._video = None

        outer = QVBoxLayout(self)
        outer.setContentsMargins(0, 0, 0, 0)
        outer.setSpacing(0)

        # The live video fills the whole widget; its device/start controls live
        # in a separate header that the MainWindow hosts in the shared top pill.
        self._header = self._build_header()

        # body: live video on top of a placeholder (placeholder shows when off)
        body = QWidget()
        self._stack = QStackedLayout(body)
        self._stack.setStackingMode(QStackedLayout.StackAll)
        self._stack.setContentsMargins(0, 0, 0, 0)

        self._placeholder = QLabel("Camera off")
        self._placeholder.setObjectName("CamPlaceholder")
        self._placeholder.setAlignment(Qt.AlignCenter)
        self._placeholder.setWordWrap(True)
        self._placeholder.setStyleSheet("background-color: #05080c;")
        self._stack.addWidget(self._placeholder)

        if _MULTIMEDIA_OK:
            self._session = QMediaCaptureSession()
            self._video = QVideoWidget()
            self._video.setStyleSheet("background-color: #05080c;")
            self._session.setVideoOutput(self._video)
            self._stack.addWidget(self._video)
            self._media_devices = QMediaDevices(self)
            self._media_devices.videoInputsChanged.connect(self._refresh_devices)

        outer.addWidget(body, 1)

        if _MULTIMEDIA_OK:
            self._refresh_devices()
        else:
            self._device.setEnabled(False)
            self._start_btn.setEnabled(False)
            self._placeholder.setText("Qt Multimedia unavailable")
        self._show_video(False)

    def header_widget(self) -> QWidget:
        """The camera's device/start controls (hosted in the shared top pill)."""
        return self._header

    def set_header_compact(self, compact: bool) -> None:
        self._cam_label.setVisible(not compact)

    # ── keep the placeholder text legible at any tile size ───────────────────
    def resizeEvent(self, e) -> None:
        super().resizeEvent(e)
        side = min(self.width(), self.height())
        px = max(8, min(20, int(side * 0.12)))
        # set via inline QSS — a per-widget stylesheet overrides the app-level
        # `QWidget { font-size }`, which a plain setFont() would not.
        self._placeholder.setStyleSheet(f"background-color: #05080c; font-size: {px}px;")

    # ── header ────────────────────────────────────────────────────────────────
    def _build_header(self) -> QWidget:
        bar = QWidget()
        bar.setAttribute(Qt.WA_TranslucentBackground, True)
        row = QHBoxLayout(bar)
        row.setContentsMargins(0, 0, 0, 0)
        row.setSpacing(6)

        self._cam_label = QLabel("CAM")
        self._cam_label.setObjectName("PanelTitle")
        row.addWidget(self._cam_label)

        self._device = QComboBox()
        self._device.setMinimumWidth(96)
        self._device.currentIndexChanged.connect(self._on_device_changed)
        row.addWidget(self._device, 1)

        refresh = QPushButton("⟳")
        refresh.setObjectName("IconButton")
        refresh.setToolTip("Refresh camera list")
        refresh.setCursor(Qt.PointingHandCursor)
        refresh.clicked.connect(self._refresh_devices)
        row.addWidget(refresh)

        self._start_btn = QPushButton("Start")
        self._start_btn.setObjectName("Ghost")
        self._start_btn.setCursor(Qt.PointingHandCursor)
        self._start_btn.clicked.connect(self._toggle)
        row.addWidget(self._start_btn)
        return bar

    # ── device handling ─────────────────────────────────────────────────────
    def _refresh_devices(self) -> None:
        if not _MULTIMEDIA_OK:
            return
        prev = self._device.currentData()
        self._device.blockSignals(True)
        self._device.clear()
        cams = QMediaDevices.videoInputs()
        for cam in cams:
            self._device.addItem(cam.description(), cam)
        self._device.blockSignals(False)
        if not cams:
            self._placeholder.setText("No camera")
            self._start_btn.setEnabled(False)
            return
        self._start_btn.setEnabled(True)
        # keep the previous selection if it's still around
        for i in range(self._device.count()):
            dev = self._device.itemData(i)
            if prev is not None and dev.id() == prev.id():
                self._device.setCurrentIndex(i)
                break
        if not self._running:
            self._placeholder.setText("Camera off")

    def _on_device_changed(self, _idx: int) -> None:
        if self._running:
            # restart on the newly selected device
            self._stop()
            self.ensure_started()

    # ── start / stop ─────────────────────────────────────────────────────────
    def _toggle(self) -> None:
        if self._running:
            self._stop()
        else:
            self.ensure_started()

    def ensure_started(self) -> bool:
        """Start the selected camera if possible. Safe to call repeatedly."""
        if not _MULTIMEDIA_OK or self._running:
            return self._running
        dev = self._device.currentData()
        if dev is None:
            return False
        try:
            self._camera = QCamera(dev)
            self._camera.errorOccurred.connect(self._on_camera_error)
            self._session.setCamera(self._camera)
            self._camera.start()
        except Exception as exc:  # pragma: no cover - hardware/driver dependent
            self._placeholder.setText(f"Camera error:\n{exc}")
            self._camera = None
            return False
        self._set_running(True)
        return True

    def _stop(self) -> None:
        if self._camera is not None:
            try:
                self._camera.stop()
            except Exception:
                pass
            self._session.setCamera(None)
            self._camera = None
        self._set_running(False)
        self._placeholder.setText("Camera off")

    def _on_camera_error(self, _err, msg: str) -> None:  # pragma: no cover
        self._placeholder.setText(f"Camera error:\n{msg or 'unavailable'}")
        self._set_running(False)

    def _set_running(self, running: bool) -> None:
        if running == self._running:
            self._show_video(running)
            return
        self._running = running
        self._start_btn.setText("Stop" if running else "Start")
        self._show_video(running)
        self.running_changed.emit(running)

    def _show_video(self, show: bool) -> None:
        if self._video is not None:
            self._video.setVisible(show)
        self._placeholder.setVisible(not show)

    def is_running(self) -> bool:
        return self._running

    # ── lifecycle ─────────────────────────────────────────────────────────────
    def shutdown(self) -> None:
        self._stop()

File: cpp/src/ui/camera_view.cpp
#include "ui/camera_view.h"

#include "domain/telemetry.h"
#include "interfaces/vision.h"
#include "ui/video_display.h"
#include "vision/centering_policy.h"
#include "vision/centroid_tracker.h"
#include "vision/motion_detector.h"
#include "vision/vision_pipeline.h"

#include <QCamera>
#include <QCameraDevice>
#include <QComboBox>
#include <QHBoxLayout>
#include <QLabel>
#include <QMediaCaptureSession>
#include <QMediaDevices>
#include <QPushButton>
#include <QStackedLayout>
#include <QVBoxLayout>
#include <QVideoFrame>
#include <QVideoSink>

namespace gcs::ui {

CameraView::CameraView(QWidget *parent) : QWidget(parent)
{
    // Pipeline thị giác: detector (placeholder AI) → tracker → policy gimbal.
    // Giữ con trỏ policy để bật/tắt gửi lệnh; quyền sở hữu chuyển vào pipeline.
    auto detector = std::make_unique<vision::MotionDetector>();
    auto tracker = std::make_unique<vision::CentroidTracker>();
    auto policy = std::make_unique<vision::CenteringPolicy>();
    m_policy = policy.get();
    m_pipeline = std::make_unique<vision::VisionPipeline>(
        std::move(detector), std::move(tracker), std::move(policy));

    auto *outer = new QVBoxLayout(this);
    outer->setContentsMargins(0, 0, 0, 0);
    outer->setSpacing(0);

    m_header = buildHeader();

    auto *bodyW = new QWidget;
    m_stack = new QStackedLayout(bodyW);
    m_stack->setStackingMode(QStackedLayout::StackAll);
    m_stack->setContentsMargins(0, 0, 0, 0);

    m_placeholder = new QLabel(QStringLiteral("Camera tắt"));
    m_placeholder->setObjectName("CamPlaceholder");
    m_placeholder->setAlignment(Qt::AlignCenter);
    m_placeholder->setWordWrap(true);
    m_placeholder->setStyleSheet(QStringLiteral("background-color: #05080c;"));
    m_stack->addWidget(m_placeholder);

    m_session = new QMediaCaptureSession(this);
    m_sink = new QVideoSink(this);
    m_session->setVideoSink(m_sink);
    connect(m_sink, &QVideoSink::videoFrameChanged, this, &CameraView::onFrame);

    m_display = new VideoDisplay;
    m_stack->addWidget(m_display);

    m_mediaDevices = new QMediaDevices(this);
    connect(m_mediaDevices, &QMediaDevices::videoInputsChanged, this, &CameraView::refreshDevices);

    outer->addWidget(bodyW, 1);

    refreshDevices();
    showVideo(false);
}

CameraView::~CameraView() = default;

void CameraView::setCommandSinkProvider(std::function<interfaces::ICommandSink *()> provider)
{
    m_sinkProvider = std::move(provider);
}

QWidget *CameraView::buildHeader()
{
    auto *bar = new QWidget;
    bar->setAttribute(Qt::WA_TranslucentBackground, true);
    auto *row = new QHBoxLayout(bar);
    row->setContentsMargins(0, 0, 0, 0);
    row->setSpacing(6);

    m_camLabel = new QLabel(QStringLiteral("CAM"));
    m_camLabel->setObjectName("PanelTitle");
    row->addWidget(m_camLabel);

    m_device = new QComboBox;
    m_device->setMinimumWidth(96);
    connect(m_device, &QComboBox::currentIndexChanged, this, &CameraView::onDeviceChanged);
    row->addWidget(m_device, 1);

    auto *refresh = new QPushButton(QStringLiteral("⟳"));
    refresh->setObjectName("IconButton");
    refresh->setToolTip(QStringLiteral("Làm mới danh sách camera"));
    refresh->setCursor(Qt::PointingHandCursor);
    connect(refresh, &QPushButton::clicked, this, &CameraView::refreshDevices);
    row->addWidget(refresh);

    m_trackBtn = new QPushButton(QStringLiteral("Bắt bám"));
    m_trackBtn->setObjectName("Ghost");
    m_trackBtn->setCursor(Qt::PointingHandCursor);
    m_trackBtn->setToolTip(QStringLiteral("Bật bám mục tiêu — lái gimbal theo người"));
    connect(m_trackBtn, &QPushButton::clicked, this, &CameraView::toggleTracking);
    row->addWidget(m_trackBtn);

    m_startBtn = new QPushButton(QStringLiteral("Bật"));
    m_startBtn->setObjectName("Ghost");
    m_startBtn->setCursor(Qt::PointingHandCursor);
    connect(m_startBtn, &QPushButton::clicked, this, &CameraView::toggle);
    row->addWidget(m_startBtn);
    return bar;
}

void CameraView::setHeaderCompact(bool compact)
{
    m_camLabel->setVisible(!compact);
}

void CameraView::resizeEvent(QResizeEvent *e)
{
    QWidget::resizeEvent(e);
    const int side = std::min(width(), height());
    const int px = std::max(8, std::min(20, int(side * 0.12)));
    m_placeholder->setStyleSheet(
        QStringLiteral("background-color: #05080c; font-size: %1px;").arg(px));
}

void CameraView::refreshDevices()
{
    QByteArray prevId;
    if (m_device->currentIndex() >= 0)
        prevId = m_device->currentData().value<QCameraDevice>().id();

    m_device->blockSignals(true);
    m_device->clear();
    const QList<QCameraDevice> cams = QMediaDevices::videoInputs();
    for (const QCameraDevice &cam : cams)
        m_device->addItem(cam.description(), QVariant::fromValue(cam));
    m_device->blockSignals(false);

    if (cams.isEmpty()) {
        m_placeholder->setText(QStringLiteral("Không có camera"));
        m_startBtn->setEnabled(false);
        return;
    }
    m_startBtn->setEnabled(true);
    for (int i = 0; i < m_device->count(); ++i) {
        if (m_device->itemData(i).value<QCameraDevice>().id() == prevId && !prevId.isEmpty()) {
            m_device->setCurrentIndex(i);
            break;
        }
    }
    if (!m_running)
        m_placeholder->setText(QStringLiteral("Camera tắt"));
}

void CameraView::onDeviceChanged(int)
{
    if (m_running) {
        stop();
        ensureStarted();
    }
}

void CameraView::toggle()
{
    if (m_running)
        stop();
    else
        ensureStarted();
}

bool CameraView::ensureStarted()
{
    if (m_running)
        return true;
    if (m_device->currentIndex() < 0)
        return false;
    const QCameraDevice dev = m_device->currentData().value<QCameraDevice>();
    if (dev.isNull())
        return false;
    m_camera = new QCamera(dev, this);
    connect(m_camera, &QCamera::errorOccurred, this, [this](QCamera::Error, const QString &msg) {
        m_placeholder->setText(QStringLiteral("Lỗi camera:\n%1")
                                   .arg(msg.isEmpty() ? QStringLiteral("không khả dụng") : msg));
        setRunning(false);
    });
    m_session->setCamera(m_camera);
    m_camera->start();
    setRunning(true);
    return true;
}

void CameraView::stop()
{
    if (m_camera) {
        m_camera->stop();
        m_session->setCamera(nullptr);
        m_camera->deleteLater();
        m_camera = nullptr;
    }
    m_pipeline->reset();
    m_display->clear();
    setRunning(false);
    m_placeholder->setText(QStringLiteral("Camera tắt"));
}

void CameraView::onFrame()
{
    if (!m_running)
        return;
    const QVideoFrame vf = m_sink->videoFrame();
    QImage img = vf.toImage();
    if (img.isNull())
        return;
    if (img.format() != QImage::Format_RGB32 && img.format() != QImage::Format_ARGB32)
        img = img.convertToFormat(QImage::Format_RGB32);

    interfaces::VideoFrame frame;
    frame.image = img;
    frame.timestampMs = domain::nowMs();

    interfaces::ICommandSink *sink = m_sinkProvider ? m_sinkProvider() : nullptr;
    const interfaces::TrackResult res = m_pipeline->process(frame, sink);

    m_display->setFrame(img);
    m_display->setResult(res);
}

void CameraView::toggleTracking()
{
    const bool on = !m_pipeline->trackingEnabled();
    m_pipeline->setTrackingEnabled(on);
    m_policy->setEnabled(on);
    m_trackBtn->setText(on ? QStringLiteral("Đang bám") : QStringLiteral("Bắt bám"));
}

void CameraView::setRunning(bool running)
{
    if (running == m_running) {
        showVideo(running);
        return;
    }
    m_running = running;
    m_startBtn->setText(running ? QStringLiteral("Tắt") : QStringLiteral("Bật"));
    showVideo(running);
    emit runningChanged(running);
}

void CameraView::showVideo(bool show)
{
    if (m_display)
        m_display->setVisible(show);
    m_placeholder->setVisible(!show);
}

void CameraView::shutdown()
{
    stop();
}

} // namespace gcs::ui
